client/notices/index.js

```diff
@@ -37,14 +37,20 @@ module.exports = {
 			container: container,
 			button: options.button,
 			href: options.href,
-			onClick: options.onClick,
 			onRemoveCallback: options.onRemoveCallback || function() {},
 			arrow: options.arrow,
 			isCompact: options.isCompact,
 			showDismiss: options.showDismiss,
 			persistent: options.persistent
 		};
 
+		if ( 'function' === typeof options.onClick ) {
+			const closeFn = this.removeNotice.bind( this, noticeObject );
+			noticeObject.onClick = ( event ) => {
+				return options.onClick( event, closeFn );
+			}
+		}
+
```

Comment:
I agree this fix makes a lot more sense. One small thing – the asymmetry of setting the option is worrying me. `onClick` is a strange exception, being set outside of the object literal. Can we do something about this?